File: cpp/tests/test_three_way.cpp
// Cross-implementation tests: exact scan, our HNSW, our HNSW with a router,
// and hnswlib.
//
// Two distinct jobs here.
//
// The first is *shared correctness*. Anything true of "a vector index" rather
// than of one particular index gets asserted for all of them: results are a
// subset of the corpus, sorted best-first, free of duplicates; k>n clamps;
// filters are honoured. Writing those once and running them over every
// implementation is what stops one of them drifting quietly.
//
// The second is *agreement*. Each index is separately graded against the exact
// oracle, but that is not sufficient — two indexes can each score 0.96 against
// the truth while overlapping far less with each other, which would mean they
// are missing different things and something is wrong. So they are also
// compared pairwise.
//
// hnswlib is a benchmark baseline only. It is here to answer whether our HNSW
// is competitive or a strawman, since a router that beats a slow baseline has
// proved nothing. All hnswlib tests compile out when it was not fetched.

#include <gtest/gtest.h>

#include <algorithm>
#include <random>
#include <set>
#include <string>
#include <vector>

#include "index/flat.hpp"
#include "index/hnsw.hpp"
#include "index/router.hpp"

#ifdef HYLIS_HAS_HNSWLIB
#include "index/hnswlib_adapter.hpp"
using hylis::index::HnswlibIndex;
#endif

using hylis::index::FlatIndex;
using hylis::index::HnswIndex;
using hylis::index::Metric;
using hylis::index::Neighbor;
using hylis::index::NeuralRouter;

namespace {

constexpr std::size_t kN = 4000;
constexpr std::size_t kDim = 16;
constexpr std::size_t kQueries = 60;
constexpr std::size_t kClusters = 32;

struct Corpus {
    std::vector<float> base, queries;
};

const Corpus& corpus() {
    static const Corpus c = [] {
        std::mt19937 rng(4242);
        std::normal_distribution<float> gauss;
        std::vector<float> centres(kClusters * kDim);
        for (float& v : centres) v = gauss(rng);
        std::uniform_int_distribution<std::size_t> pick(0, kClusters - 1);

        Corpus out{std::vector<float>(kN * kDim), std::vector<float>(kQueries * kDim)};
        for (std::size_t i = 0; i < kN; ++i) {
            const std::size_t k = pick(rng);
            for (std::size_t j = 0; j < kDim; ++j) {
                out.base[i * kDim + j] = centres[k * kDim + j] + 0.15f * gauss(rng);
            }
        }
        for (std::size_t i = 0; i < kQueries; ++i) {
            const std::size_t k = pick(rng);
            for (std::size_t j = 0; j < kDim; ++j) {
                out.queries[i * kDim + j] = centres[k * kDim + j] + 0.15f * gauss(rng);
            }
        }
        return out;
    }();
    return c;
}

std::vector<std::int64_t> ids_of(const std::vector<Neighbor>& ns) {
    std::vector<std::int64_t> out;
    for (const Neighbor& n : ns) out.push_back(n.id);
    return out;
}

// A router built directly from the corpus: cluster centres are just evenly
// spaced base vectors, and the first layer projects onto them so the predicted
// cluster is the nearest centre. Crude next to the trained one, but it makes
// these tests independent of the Python trainer.
NeuralRouter geometric_router() {
    const Corpus& c = corpus();
    const std::size_t clusters = 32;
    const std::size_t stride = kN / clusters;

    std::string w1, b1, w2, b2, medoids;
    // hidden == clusters: hidden unit j scores the query against centre j,
    // as -(|centre|^2/2 - q.centre), then w2 is the identity.
    for (std::size_t i = 0; i < kDim; ++i) {
        for (std::size_t j = 0; j < clusters; ++j) {
            if (!w1.empty()) w1 += ",";
            w1 += std::to_string(c.base[(j * stride) * kDim + i]);
        }
    }
    for (std::size_t j = 0; j < clusters; ++j) {
        double sq = 0.0;
        for (std::size_t i = 0; i < kDim; ++i) {
            const double v = c.base[(j * stride) * kDim + i];
            sq += v * v;
        }
        if (!b1.empty()) b1 += ",";
        b1 += std::to_string(-0.5 * sq);
        if (!b2.empty()) b2 += ",";
        b2 += "0";
        if (!medoids.empty()) medoids += ",";
        medoids += std::to_string(j * stride);
    }
    for (std::size_t h = 0; h < clusters; ++h) {
        for (std::size_t k = 0; k < clusters; ++k) {
            if (!w2.empty()) w2 += ",";
            w2 += (h == k ? "1" : "0");
        }
    }

    return NeuralRouter::from_json(
        "{\"dim\":" + std::to_string(kDim) + ",\"hidden\":" + std::to_string(clusters) +
        ",\"clusters\":" + std::to_string(clusters) + ",\"w1\":[" + w1 + "],\"b1\":[" +
        b1 + "],\"w2\":[" + w2 + "],\"b2\":[" + b2 + "],\"medoids\":[" + medoids + "]}");
}

// Mean fraction of `truth` that `found` recovered, compared as sets per query.
double overlap(const std::vector<std::vector<std::int64_t>>& found,
               const std::vector<std::vector<std::int64_t>>& truth, std::size_t k) {
    std::size_t hits = 0;
    for (std::size_t i = 0; i < found.size(); ++i) {
        const std::set<std::int64_t> want(truth[i].begin(), truth[i].end());
        for (std::int64_t id : found[i]) {
            if (want.count(id)) ++hits;
        }
    }
    return static_cast<double>(hits) / static_cast<double>(found.size() * k);
}

// Every implementation answers through this, so the shared assertions below
// can be written once.
using Searcher = std::function<std::vector<Neighbor>(const float*, std::size_t)>;

std::vector<std::pair<std::string, Searcher>> implementations() {
    static FlatIndex flat(kDim);
    static HnswIndex graph(kDim);
    static HnswIndex routed(kDim);
    static bool built = false;
    if (!built) {
        const Corpus& c = corpus();
        flat.add_batch(c.base.data(), kN);
        graph.add_batch(c.base.data(), kN);
        routed.add_batch(c.base.data(), kN);
        routed.set_router(geometric_router());
        built = true;
    }

    std::vector<std::pair<std::string, Searcher>> out;
    out.emplace_back("flat", [](const float* q, std::size_t k) {
        return flat.search(q, k);
    });
    out.emplace_back("hnsw", [](const float* q, std::size_t k) {
        return graph.search(q, k, 100);
    });
    out.emplace_back("routed", [](const float* q, std::size_t k) {
        return routed.search(q, k, 100, /*use_router=*/true);
    });
#ifdef HYLIS_HAS_HNSWLIB
    static HnswlibIndex lib(kDim, Metric::L2, kN);
    static bool lib_built = false;
    if (!lib_built) {
        lib.add_batch(corpus().base.data(), kN);
        lib_built = true;
    }
    out.emplace_back("hnswlib", [](const float* q, std::size_t k) {
        return lib.search(q, k, 100);
    });
#endif
    return out;
}

}  // namespace

// --------------------------------------------------------------------------
// Shared correctness: true of any vector index, asserted for all of them
// --------------------------------------------------------------------------

TEST(ThreeWay, ResultsAreValidSortedAndDistinct) {
    const Corpus& c = corpus();
    for (const auto& [name, search] : implementations()) {
        for (std::size_t i = 0; i < kQueries; ++i) {
            const auto found = search(c.queries.data() + i * kDim, 10);
            ASSERT_EQ(found.size(), 10u) << name << " query " << i;

            std::set<std::int64_t> seen;
            for (std::size_t j = 0; j < found.size(); ++j) {
                EXPECT_GE(found[j].id, 0) << name;
                EXPECT_LT(found[j].id, static_cast<std::int64_t>(kN)) << name;
                EXPECT_TRUE(seen.insert(found[j].id).second)
                    << name << " returned a duplicate id";
                if (j) {
                    EXPECT_LE(found[j - 1].score, found[j].score)
                        << name << " results are not sorted best-first";
                }
            }
        }
    }
}

TEST(ThreeWay, KLargerThanTheCorpusIsClamped) {
    const Corpus& c = corpus();
    for (const auto& [name, search] : implementations()) {
        EXPECT_LE(search(c.queries.data(), kN + 500).size(), kN) << name;
    }
}

TEST(ThreeWay, ZeroKReturnsNothing) {
    const Corpus& c = corpus();
    for (const auto& [name, search] : implementations()) {
        EXPECT_TRUE(search(c.queries.data(), 0).empty()) << name;
    }
}

TEST(ThreeWay, EveryImplementationClearsTheRecallFloor) {
    const Corpus& c = corpus();
    static FlatIndex exact(kDim);
    static bool built = false;
    if (!built) { exact.add_batch(c.base.data(), kN); built = true; }

    std::vector<std::vector<std::int64_t>> truth;
    for (std::size_t i = 0; i < kQueries; ++i) {
        truth.push_back(ids_of(exact.search(c.queries.data() + i * kDim, 10)));
    }

    for (const auto& [name, search] : implementations()) {
        std::vector<std::vector<std::int64_t>> got;
        for (std::size_t i = 0; i < kQueries; ++i) {
            got.push_back(ids_of(search(c.queries.data() + i * kDim, 10)));
        }
        EXPECT_GE(overlap(got, truth, 10), 0.90)
            << name << " fell below the recall floor against the exact oracle";
    }
}

// --------------------------------------------------------------------------
// Agreement: each against the oracle is not enough
// --------------------------------------------------------------------------

TEST(ThreeWay, ImplementationsAgreeWithEachOtherNotJustWithTheOracle) {
    const Corpus& c = corpus();
    const auto impls = implementations();

    std::vector<std::vector<std::vector<std::int64_t>>> results;
    for (const auto& [name, search] : impls) {
        std::vector<std::vector<std::int64_t>> got;
        for (std::size_t i = 0; i < kQueries; ++i) {
            got.push_back(ids_of(search(c.queries.data() + i * kDim, 10)));
        }
        results.push_back(std::move(got));
    }

    for (std::size_t a = 0; a < impls.size(); ++a) {
        for (std::size_t b = a + 1; b < impls.size(); ++b) {
            EXPECT_GE(overlap(results[a], results[b], 10), 0.85)
                << impls[a].first << " and " << impls[b].first
                << " each look accurate but disagree with one another, which "
                   "means they are missing different neighbours";
        }
    }
}

TEST(ThreeWay, EveryImplementationRetrievesAStoredVectorItself) {
    const Corpus& c = corpus();
    for (const auto& [name, search] : implementations()) {
        std::size_t hits = 0, probes = 0;
        for (std::size_t i = 0; i < kN; i += 137) {
            ++probes;
            const auto found = search(c.base.data() + i * kDim, 1);
            if (!found.empty() && found[0].id == static_cast<std::int64_t>(i)) ++hits;
        }
        EXPECT_GT(hits * 100, probes * 95) << name << " failed self-retrieval";
    }
}

// --------------------------------------------------------------------------
// hnswlib, the external reference
// --------------------------------------------------------------------------

#ifdef HYLIS_HAS_HNSWLIB

TEST(HnswlibBaseline, ScoresUseOurConventionNotHnswlibs) {
    // hnswlib returns squared L2; we return true distance. If the adapter did
    // not convert, every score comparison against it would silently be
    // comparing different quantities.
    const Corpus& c = corpus();
    FlatIndex exact(kDim);
    exact.add_batch(c.base.data(), kN);
    HnswlibIndex lib(kDim, Metric::L2, kN);
    lib.add_batch(c.base.data(), kN);

    const auto truth = exact.search(c.queries.data(), 1);
    const auto got = lib.search(c.queries.data(), 1, 200);
    ASSERT_FALSE(got.empty());
    ASSERT_FALSE(truth.empty());
    if (got[0].id == truth[0].id) {
        EXPECT_NEAR(got[0].score, truth[0].score, 1e-3f);
    }
}

TEST(HnswlibBaseline, OurImplementationIsCompetitive) {
    // The question this baseline exists to answer. Not "are we faster" — we
    // are a hand-written implementation against a heavily optimised one — but
    // "is our recall in the same league", because a router that beats a
    // strawman has proved nothing.
    const Corpus& c = corpus();
    FlatIndex exact(kDim);
    exact.add_batch(c.base.data(), kN);
    HnswIndex ours(kDim);
    ours.add_batch(c.base.data(), kN);
    HnswlibIndex lib(kDim, Metric::L2, kN);
    lib.add_batch(c.base.data(), kN);

    std::vector<std::vector<std::int64_t>> truth, mine, theirs;
    for (std::size_t i = 0; i < kQueries; ++i) {
        const float* q = c.queries.data() + i * kDim;
        truth.push_back(ids_of(exact.search(q, 10)));
        mine.push_back(ids_of(ours.search(q, 10, 50)));
        theirs.push_back(ids_of(lib.search(q, 10, 50)));
    }

    const double ours_recall = overlap(mine, truth, 10);
    const double lib_recall = overlap(theirs, truth, 10);
    EXPECT_GT(ours_recall, lib_recall - 0.05)
        << "ours " << ours_recall << " vs hnswlib " << lib_recall
        << " — more than five points behind the reference suggests a defect, "
           "not just a tuning difference";
}

TEST(HnswlibBaseline, HonoursFilters) {
    const Corpus& c = corpus();
    HnswlibIndex lib(kDim, Metric::L2, kN);
    lib.add_batch(c.base.data(), kN);

    std::vector<std::int64_t> allowed;
    for (std::int64_t i = 0; i < static_cast<std::int64_t>(kN); i += 7) {
        allowed.push_back(i);
    }
    for (std::size_t i = 0; i < 10; ++i) {
        for (const Neighbor& n : lib.search_filtered(c.queries.data() + i * kDim,
                                                     10, allowed, 200)) {
            EXPECT_EQ(n.id % 7, 0);
        }
    }
    EXPECT_THROW(lib.search_filtered(c.queries.data(), 5, {999999}), std::out_of_range);
}

#endif  // HYLIS_HAS_HNSWLIB

File: tests/test_three_way.py
"""Correctness and agreement across every vector index in the project.

Four implementations answer the same questions: the exact scan, hnswlib, our
HNSW, and our HNSW with the neural router. Two things are checked.

*Shared correctness* — anything true of "a vector index" rather than of one
particular index is asserted for all of them, once, from one place. That is
what stops one implementation drifting quietly while the others stay honest.

*Agreement* — each is graded against the exact oracle, but that alone is not
enough: two indexes can each score 0.95 against the truth while overlapping far
less with each other, which would mean they are missing different neighbours.
So they are compared pairwise as well.

hnswlib is a benchmark baseline only; these tests skip it when it was not
fetched.
"""

import numpy as np
import pytest

from hylis import FlatIndex, HnswIndex, Metric
from hylis import datasets as ds
from hylis._hnsw import NeuralRouter
from hylis.router import kmeans, build_training_set, RouterMLP, router_json

import hylis._hnswlib as hnswlib_module

HAS_HNSWLIB = getattr(hnswlib_module, "available", False)

K = 10
EF = 100


@pytest.fixture(scope="module")
def corpus():
    return ds.random_vectors(n=4000, dim=24, n_queries=80, seed=1, n_clusters=40)


@pytest.fixture(scope="module")
def truth(corpus):
    return ds.compute_ground_truth(corpus.base, corpus.queries, k=K)


@pytest.fixture(scope="module")
def indexes(corpus):
    """Every implementation, behind one uniform `search(query, k) -> ids`."""
    built = {}

    flat = FlatIndex(corpus.dim)
    flat.add_batch(corpus.base)
    built["flat"] = lambda q, k: [n.id for n in flat.search(q, k)]

    graph = HnswIndex(corpus.dim, Metric.L2, M=16, ef_construction=200)
    graph.add_batch(corpus.base)
    built["hnsw"] = lambda q, k: [n.id for n in graph.search(q, k, EF)]

    km = kmeans(corpus.base, n_clusters=48, seed=0)
    ts = build_training_set(corpus.base, km.assignment, n_samples=20000, seed=0)
    model = RouterMLP(dim=corpus.dim, clusters=48, hidden=48, seed=0)
    model.fit(ts.x_train, ts.y_train, epochs=25)

    routed = HnswIndex(corpus.dim, Metric.L2, M=16, ef_construction=200)
    routed.add_batch(corpus.base)
    routed.set_router(NeuralRouter.from_json(router_json(model, km.medoids)))
    built["routed"] = lambda q, k: [n.id for n in routed.search(q, k, EF, True)]

    if HAS_HNSWLIB:
        lib = hnswlib_module.HnswlibIndex(corpus.dim, Metric.L2, capacity=corpus.n,
                                          M=16, ef_construction=200)
        lib.add_batch(corpus.base)
        built["hnswlib"] = lambda q, k: [n.id for n in lib.search(q, k, EF)]

    return built


def results_for(search, corpus, k=K):
    return [search(q, k) for q in corpus.queries]


def overlap(a, b, k=K):
    """Mean fraction of b's ids that a also returned, per query."""
    hits = sum(len(set(x) & set(y)) for x, y in zip(a, b))
    return hits / (len(a) * k)


# --------------------------------------------------------------------------
# Shared correctness
# --------------------------------------------------------------------------


def test_every_implementation_is_present(indexes):
    assert {"flat", "hnsw", "routed"} <= set(indexes)
    if HAS_HNSWLIB:
        assert "hnswlib" in indexes


def test_results_are_valid_sorted_and_distinct(indexes, corpus):
    for name, search in indexes.items():
        for i, query in enumerate(corpus.queries[:20]):
            ids = search(query, K)
            assert len(ids) == K, f"{name} query {i} returned {len(ids)}"
            assert len(set(ids)) == K, f"{name} returned duplicates"
            assert all(0 <= j < corpus.n for j in ids), f"{name} returned a bad id"


def test_k_larger_than_the_corpus_is_clamped(indexes, corpus):
    for name, search in indexes.items():
        assert len(search(corpus.queries[0], corpus.n + 500)) <= corpus.n, name


def test_every_implementation_clears_the_recall_floor(indexes, corpus, truth):
    truth_ids = [row.tolist() for row in truth]
    for name, search in indexes.items():
        recall = overlap(results_for(search, corpus), truth_ids)
        assert recall >= 0.90, f"{name} recall {recall:.4f} against the exact oracle"


def test_self_retrieval(indexes, corpus):
    probes = list(range(0, corpus.n, 173))
    for name, search in indexes.items():
        hits = sum(1 for i in probes if search(corpus.base[i], 1)[:1] == [i])
        assert hits >= 0.95 * len(probes), f"{name} failed self-retrieval"


# --------------------------------------------------------------------------
# Agreement
# --------------------------------------------------------------------------


def test_implementations_agree_with_each_other(indexes, corpus):
    """Each matching the oracle is not sufficient — they must also match each
    other, or they are missing different neighbours."""
    results = {name: results_for(search, corpus) for name, search in indexes.items()}
    names = sorted(results)

    for i, a in enumerate(names):
        for b in names[i + 1:]:
            score = overlap(results[a], results[b])
            assert score >= 0.85, (
                f"{a} and {b} each look accurate but agree only {score:.3f} "
                "with one another"
            )


def test_agreement_matrix_is_symmetric_and_complete(indexes, corpus):
    results = {name: results_for(search, corpus) for name, search in indexes.items()}
    for a in results:
        assert overlap(results[a], results[a]) == pytest.approx(1.0)


# --------------------------------------------------------------------------
# hnswlib, the external reference
# --------------------------------------------------------------------------


@pytest.mark.skipif(not HAS_HNSWLIB, reason="hnswlib was not fetched")
def test_our_hnsw_is_competitive_with_the_reference(indexes, corpus, truth):
    """The question this baseline exists to answer.

    Not "are we faster" — we are a hand-written implementation against a
    heavily optimised one — but "is our recall in the same league", because a
    router that beats a strawman has proved nothing.
    """
    truth_ids = [row.tolist() for row in truth]
    ours = overlap(results_for(indexes["hnsw"], corpus), truth_ids)
    theirs = overlap(results_for(indexes["hnswlib"], corpus), truth_ids)
    assert ours > theirs - 0.05, (
        f"ours {ours:.4f} vs hnswlib {theirs:.4f} — more than five points "
        "behind the reference suggests a defect, not a tuning difference"
    )


@pytest.mark.skipif(not HAS_HNSWLIB, reason="hnswlib was not fetched")
def test_hnswlib_scores_use_our_convention(corpus):
    """hnswlib returns squared L2; we return true distance. If the adapter did
    not convert, score comparisons would silently compare different things."""
    exact = FlatIndex(corpus.dim)
    exact.add_batch(corpus.base)
    lib = hnswlib_module.HnswlibIndex(corpus.dim, Metric.L2, capacity=corpus.n)
    lib.add_batch(corpus.base)

    query = corpus.queries[0]
    want = exact.search(query, 1)[0]
    got = lib.search(query, 1, 200)[0]
    if got.id == want.id:
        assert got.score == pytest.approx(want.score, rel=1e-3)


@pytest.mark.skipif(not HAS_HNSWLIB, reason="hnswlib was not fetched")
def test_hnswlib_honours_filters(corpus):
    lib = hnswlib_module.HnswlibIndex(corpus.dim, Metric.L2, capacity=corpus.n)
    lib.add_batch(corpus.base)
    allowed = list(range(0, corpus.n, 7))

    for query in corpus.queries[:5]:
        for n in lib.search_filtered(query, 10, allowed, 200):
            assert n.id % 7 == 0
    with pytest.raises(IndexError):
        lib.search_filtered(corpus.queries[0], 5, [999999])


def test_module_reports_availability_either_way():
    """`import hylis._hnswlib` must always work, so an absent baseline shows up
    as a flag rather than an ImportError three frames away."""
    assert isinstance(hnswlib_module.available, bool)
